src/SAML2/XML/AbstractXMLElement.php

```diff
@@ -43,7 +43,8 @@ public function __toString(): string
      */
     public function serialize(): string
     {
-        return $this->toXML()->ownerDocument->saveXML();
+        $xml = $this->toXML();
+        return is_null($xml) ? '' : $xml->ownerDocument->saveXML();
```

Comment:
This won't work, and it might be a reason to actually return an empty element rather than `null`. The problem is that `unserialize()` relies on proper XML to reconstruct the object (via each object's `fromXML()` implementation). Take `AuthnContext` in this PR as an example. `toXML()` might return `null`. In that case, serializing leads to a string that contains no XML as its contents. When trying to unserialize later, `unserialize()` will call `fromXML()`, which requires a `DOMElement` passed as a parameter. This will fail with an `InvalidArgumentException` because the argument passed to `unserialize()` (`$serialized`, later in this file) will be an empty string, [which is explicitly disallowed](https://github.com/simplesamlphp/saml2/blob/master/src/SAML2/DOMDocumentFactory.php#L30-L32).

Reply:
I see.. What about using '<root />' instead of the empty string.. Would that help?

Reply:
I don't think that would make any sense, to be honest. In any case, it would also fail since when processed into a `DOMElement`, the corresponding `fromXML()` method will check for the name of the tag, which will not match.

Reply:
OK, changing this back to non-nullable again

Reply:
Sorry!

Reply:
No worries! It's a puzzle we have to solve together..

I wonder if we can think of a clever way to determine whether the result of `$x->toXML($parent)` would be an empty tag.. I'm thinking along the lines of an isEmptyTag-method on our abstract class that will return false by default but can be overriden by the descending classes that would currently return `null`...

That would A) allow us to remove empty tags from message and B) still allow developers to work with empty tags as they desire.. What do you reckon?

Reply:
I was also thinking about that and I believe I have a solution. The thing is, it will only be some classes creating an empty tag, and in most cases we know if that'll be the case or not because we've created the object ourselves and know that we passed actual parameters to the constructor, and therefore the tag won't be empty.

For those cases were we can't know (e.g. the object was passed to us), there is actually a way to check if `toXML()` returned an empty tag:

```php
$e = $object->toXML();
if ($e->attributes->length === 0 && !$e->hasChildNodes()) {
    // empty tag
}
```

If we see this code repeating a lot, we could indeed create a simple static `isEmptyTag(DOMElement $e)` method in `AbstractXMLElement` or even `SAML2\Utils` so that we can just call that. What do you think?

Reply:
I think we should go with a simple method and it doesn't have to be static.
``
if (!$object->isEmptyTag()) {
    $object->toXML($parent);
}
``
The isEmptyTag-method can be as easy as:
`return empty($this->optional1) && empty($this->optional2);` (or `return false;` as the default case)

Reply:
Oh, so you are thinking about checking this *before* actually calling `toXML()`. That's another option, although I don't really see much benefit (except for the save of a few CPU cycles creating the empty `DOMElement`). On the contrary, the downside I see with that approach is that those classes that can produce an empty element need to implement an extra method, while the post-`toXML()` approach allows you to do the same without extra code.

Reply:
Agreed, we'll go with your solution